crates/astria-sequencer-relayer/src/data_availability.rs

```diff
@@ -352,128 +309,143 @@ impl CelestiaClient {
-                continue 'namespaces;
-            };
-            rollup_txs_map.insert(*rollup_namespace, rollup_txs);
-        }
+        let verification_key = VerificationKey::try_from(&*namespace_data.public_key)
+            .wrap_err("failed constructing verification key from stored bytes")?;
 
-        Ok(SequencerBlockData {
-            block_hash: data.block_hash.clone(),
-            header: data.header.clone(),
-            last_commit: data.last_commit.clone(),
-            rollup_txs: rollup_txs_map,
-        })
-    }
-
-    async fn get_rollup_data_for_block(
-        &self,
-        block_hash: &[u8],
-        rollup_namespace: Namespace,
-        height: u64,
-        verification_key: Option<VerificationKey>,
-    ) -> eyre::Result<Option<Vec<IndexedTransaction>>> {
+        let namespace_ids = namespace_data
+            .data
+            .rollup_namespaces
+            .iter()
+            .map(|ns| **ns)
+            .collect();
         let req = GetAllRequest {
             height,
-            namespace_ids: vec![*rollup_namespace],
+            namespace_ids,
         };
         let rsp = self
             .client
             .blob_get_all(req)
             .await
             .wrap_err("failed getting namespaced data")?;
 
-        let mut rollup_datas = rsp
+        // get only rollup datas that can be deserialized
+        let mut rollup_datas: HashMap<_, _> = rsp
             .blobs
             .iter()
             .filter_map(|blob| {
                 if let Ok(data) = SignedNamespaceData::<RollupNamespaceData>::from_bytes(&blob.data)
                 {
-                    Some(data)
+                    Some((Namespace::new(blob.namespace_id), data))
                 } else {
+                    // FIXME: provide some info to identify the rollup namespace data?
                     warn!("failed to deserialize rollup namespace data");
                     None
                 }
             })
-            .filter(|d| {
-                let hash = match d.data.hash() {
-                    Ok(hash) => hash,
-                    Err(_) => return false,
-                };
-
-                match Signature::try_from(&*d.signature) {
-                    Ok(sig) => {
-                        let Some(verification_key) = verification_key else {
-                            return true;
-                        };
-                        verification_key.verify(&sig, &hash).is_ok()
-                    }
-                    Err(_) => false,
-                }
-            })
-            .filter(|d| d.data.block_hash == block_hash);
+            .collect();
+        // retain rollup datas whose block hash matches the block hash of the namespaced data
+        rollup_datas
+            .retain(|_, rollup_data| namespace_data.data.block_hash == rollup_data.data.block_hash);
+
+        // retain rollup datas that can be verified
+        rollup_datas.retain(|namespace, rollup_data| {
```

Comment:
maybe add a `retain` before this that checks the public key which signed the rollup data is the same as `verification_key`? might save on some serialization/hashing

Reply:
I am leaving this to a followup PR because I think this might require some rethinking of our types. It's all a bit confusing right now.